doc(adding_interactivity): translate in french the index (#468)
reactjs/fr.react.dev: src/content/learn/adding-interactivity.md

---
title: Ajout d'interactivité
---

<Intro>

Certaines choses à l'écran se mettent à jour en réponse aux entrées de l'utilisateur. Par exemple, en cliquant sur une galerie d'images, l'image active change. En React, les données qui changent au fil du temps sont appelées *états.* Vous pouvez ajouter un état à n'importe quel composant et le mettre à jour au besoin. Dans ce chapitre, vous apprendrez à écrire des composants qui gèrent les interactions, mettent à jour leur état et affichent un visuel différent au fil du temps.

</Intro>

<YouWillLearn isChapter={true}>

* [Comment gérer les événements initiés par l'utilisateur](/learn/responding-to-events)
* [Comment faire en sorte que les composants "se souviennent" des informations en utilisant des états](/learn/state-a-components-memory)
* [Comment React met à jour l'interface utilisateur en deux phases](/learn/render-and-commit)
* [Pourquoi l'état ne se met pas à jour immédiatement après sa modification](/learn/state-as-a-snapshot)

> * [Comment mettre en file d'attente plusieurs mises à jour de l'état](/learn/queueing-a-series-of-state-updates)

Reviewer: ```suggestion
* [Comment cumuler plusieurs mises à jour d’un même état](/learn/queueing-a-series-of-state-updates)
```

cf. remarque sidebar

export const sculptureList = [{
  name: 'Homenaje a la Neurocirugía',
  artist: 'Marta Colvin Andrade',

>   description: 'Bien que Colvin soit principalement connue pour ses thèmes abstraits qui font allusion à des symboles préhispaniques, cette sculpture gigantesque, un hommage à la neurochirurgie, est l\'une de ses pièces d\'art public les plus reconnaissables.',

Author: Effectivement, j'ai recoupé avec Wikipedia VF quand c'était possible :sweat_smile:

<LearnMore path="/learn/state-a-components-memory">

Lisez **[État : la mémoire d'un composant](/learn/state-a-components-memory)** pour apprendre comment mémoriser une valeur et la mettre à jour lors d'une interaction.

</LearnMore>

> ## Le rendu et la mise à jour {/*render-and-commit*/}

Reviewer: ```suggestion
## Rendu et Commit {/*render-and-commit*/}
```

cf. remarque sidebar

Avant que vos composants ne soient affichés à l'écran, ils doivent être rendus par React. Comprendre les étapes de ce processus vous aidera à réfléchir à la manière dont votre code s'exécute et à expliquer son comportement.

> Imaginez que vos composants soient des cuisiniers dans un restaurant, assemblant des plats savoureux à partir d'ingrédients. Dans ce scénario, React est le serveur qui prend les commandes des clients et leur apporte leurs plats. Ce processus de demander et de servir de l'interface utilisateur comporte trois étapes :

Reviewer: ```suggestion
Imaginez que vos composants soient des cuisiniers dans un restaurant, assemblant des plats savoureux à partir d'ingrédients. Dans ce scénario, React est le serveur qui prend les commandes des clients et leur apporte leurs plats. Ce processus de demande et de service de l'UI comporte trois étapes :
```

On évite autant que possible les subordonnées ou groupes nominaux basés sur un infinitif, c'est ampoulé.

> 1. **Déclencher** un rendu (livrer la commande du client à la cuisine)

Reviewer: ```suggestion
1. **Déclencher** un rendu (envoyer la commande du client à la cuisine)
```

"livrer" est trop connoté dans l'autre sens

2. **Faire le rendu** du composant (préparer la commande en cuisine)

> 3. **Mettre à jour** le DOM (placer la commande sur la table du client)

Reviewer: ```suggestion
3. **Mettre à jour** le DOM (phase de Commit ;  revient à déposer la commande sur la table du client)
```

<IllustrationBlock sequential>
  <Illustration caption="Déclencher" alt="React agit comme un serveur dans un restaurant, qui récupère les commandes des utilisateurs et les transmet à la cuisine des composants." src="/images/docs/illustrations/i_render-and-commit1.png" />
  <Illustration caption="Faire le rendu" alt="Le chef Card fournit à React un nouveau composant Card." src="/images/docs/illustrations/i_render-and-commit2.png" />

>   <Illustration caption="Mettre à jour" alt="React livre Card à l'utilisateur à sa table." src="/images/docs/illustrations/i_render-and-commit3.png" />

Reviewer: ```suggestion
  <Illustration caption="Mettre à jour (Commit)" alt="React dépose le Card sur la table de l’utilisateur." src="/images/docs/illustrations/i_render-and-commit3.png" />
```

</IllustrationBlock>

<LearnMore path="/learn/render-and-commit">

> Lisez **[Le rendu et la mise à jour](/learn/render-and-commit)** pour apprendre sur le cycle de vie de la mise à jour d'une interface utilisateur.

Reviewer: ```suggestion
Lisez **[Rendu et Commit](/learn/render-and-commit)** pour apprendre sur le cycle de vie d’une mise à jour de l’interface.
```

Léger glissement sémantique. Par ailleurs, on entrelace "UI" et "interface" selon l'humeur au lieu du loooong "interface utilisateur", qu'on a relié à l'abréviation en début de page.

</LearnMore>

> ## L'état comme une capture {/*state-as-a-snapshot*/}

Reviewer: ```suggestion
## L'état comme un instantané {/*state-as-a-snapshot*/}
```

cf. remarque sidebar

> Contrairement aux variables JavaScript classiques, une variable d'état de React se comporte plus comme une capture à un instant T. Lui affecter une nouvelle valeur ne change pas la variable d'état que vous avez déjà, mais déclenche plutôt un nouveau rendu. Cela peut être surprenant au début !

Reviewer: ```suggestion
Contrairement aux variables JavaScript classiques, une variable d'état dans React se comporte davantage comme une photo instantanée. Lui affecter une nouvelle valeur ne change pas la variable d'état que vous avez déjà, mais déclenche plutôt un nouveau rendu. Ça peut surprendre au début !
```

```js
console.log(count);  // 0

> setCount(count + 1); // Déclenche un nouveau rendu avec la valeur 1

Reviewer: ```suggestion
setCount(count + 1); // Entraînera un nouveau rendu avec la valeur 1
```

> console.log(count);  // Toujours 0!

Reviewer: ```suggestion
console.log(count);  // Toujours 0 !
```

```

> Ce comportement vous aide à éviter des bugs subtils. Voici une petite application de chat. Essayez de deviner ce qui se passe si vous appuyez sur "Envoyer" d'abord, *puis* changez le destinataire en Bob. Quel nom apparaîtra dans `alert` cinq secondes plus tard ?

Reviewer: ```suggestion
Ce comportement vous aide à éviter des bugs subtils. Voici une petite appli de discussion. Essayez de deviner ce qui se passe si vous appuyez sur « Envoyer » d'abord, *et ensuite* changez le destinataire pour Bob. Quel nom apparaîtra dans le `alert` cinq secondes plus tard ?
```

  function handleSubmit(e) {
    e.preventDefault();
    setTimeout(() => {
      alert(`Vouz avez dit ${message} à ${to}`);
    }, 5000);
  }

  return (
    <form onSubmit={handleSubmit}>
      <label>

>         À:{' '}

Reviewer: ```suggestion
        À :{' '}
```

<LearnMore path="/learn/state-as-a-snapshot">

> Lisez **[L'état comme une capture](/learn/state-as-a-snapshot)** pour comprendre pourquoi un état semble "fixe" et immuable à l'intérieur des gestionnaires d'événements.

Reviewer: ```suggestion
Lisez **[L'état comme un instantané](/learn/state-as-a-snapshot)** pour comprendre pourquoi un état semble « fixe » et immuable à l’intérieur des gestionnaires d’événements.
```

</Sandpack>

> [L'état comme une capture](/learn/state-as-a-snapshot) explique pourquoi cela se produit. Affecter une nouvelle valeur à un état déclenche un nouveau rendu, mais ne change pas sa valeur dans le code en cours d'exécution. Ainsi, `score` reste à `0` juste après avoir appelé `setScore(score + 1)`.

Reviewer: ```suggestion
[L'état comme un instantané](/learn/state-as-a-snapshot) en explique la raison. Affecter une nouvelle valeur à un état déclenchera un nouveau rendu, mais ne change pas sa valeur dans le code en cours d’exécution. Ainsi, `score` reste à `0` juste après avoir appelé `setScore(score + 1)`.
```

console.log(score);  // 0
```

> Vous pouvez corriger cela en passant une *fonction de mise à jour* lorsque vous affectez une nouvelle valeur à l'état. Remarquez comment le remplacement de `setScore(score + 1)` par `setScore(s => s + 1)` corrige le bouton "+3". Cela vous permet de mettre en file d'attente plusieurs mises à jour d'état.

Reviewer: ```suggestion
Vous pouvez corriger ça en passant une *fonction de mise à jour* lorsque vous affectez une nouvelle valeur à l’état. Voyez comme le remplacement de `setScore(score + 1)` par `setScore(s => s + 1)` corrige le bouton "+3". Ça vous permet de cumuler plusieurs mises à jour d’un même état.
```

<LearnMore path="/learn/queueing-a-series-of-state-updates">

> Lisez **[Mettre en file d'attente une série de mises à jour d'état](/learn/queueing-a-series-of-state-updates)** pour apprendre comment mettre en file d'attente une séquence de mises à jour d'état.

Reviewer: ```suggestion
Lisez **[Cumuler les série de mises à jour d’un même état](/learn/queueing-a-series-of-state-updates)** pour apprendre comment cumuler plusieurs mises à jour d’une même variable d’état.
```

</LearnMore>

## Mettre à jour les objets d'un état {/*updating-objects-in-state*/}

Un état peut contenir n'importe quel type de valeur JavaScript, y compris des objets. Cependant, vous ne devez pas changer directement les objets et les tableaux que vous stockez dans l'état React. Au lieu de cela, lorsque vous voulez mettre à jour un objet ou un tableau, vous devez en créer un nouveau (ou faire une copie de l'existant), puis mettre à jour l'état pour utiliser cette copie.

> Généralement, vous utiliserez la syntaxe de décomposition `...` pour copier les objets et les tableaux que vous souhaitez modifier. Par exemple, la mise à jour d'un objet imbriqué pourrait ressembler à ceci :

Reviewer: ```suggestion
Généralement, vous utiliserez la syntaxe de *spread* `...` pour copier les objets et les tableaux que vous souhaitez modifier. Par exemple, la mise à jour d'un objet imbriqué pourrait ressembler à ceci :
```

"décomposition" est sans doute mon seul vrai désaccord avec Julien "Sphinx", le principal traducteur FR du MDN 😉.  D'autant qu'il utilise ce terme pour le *spread* **et** pour la déstructuration.  Ce sont deux notions distinctes, et les termes "rest / spread" sont bien installés en usage francophone.

</Sandpack>

> Si la copie d'objets dans le code devient fastidieuse, vous pouvez utiliser une bibliothèque telle que [Immer](https://github.com/immerjs/use-immer) pour réduire le code répétitif :

Reviewer: ```suggestion
Si la copie d'objets dans le code devient fastidieuse, vous pouvez utiliser une bibliothèque telle que [Immer](https://github.com/immerjs/use-immer) pour simplifier le code :
```

</Sandpack>

> Si la copie de tableaux dans le code devient fastidieuse, vous pouvez utiliser une bibliothèque telle que [Immer](https://github.com/immerjs/use-immer) pour réduire le code répétitif :

Reviewer: ```suggestion
Si la copie de tableaux dans le code devient fastidieuse, vous pouvez utiliser une bibliothèque telle que [Immer](https://github.com/immerjs/use-immer) pour simplifier le code :
```

<LearnMore path="/learn/updating-arrays-in-state">

Lisez **[Mettre à jour les tableaux d'un état](/learn/updating-arrays-in-state)** pour apprendre comment mettre à jour les tableaux correctement.

</LearnMore>

## Et maintenant ? {/*whats-next*/}

> Allez sur [Répondre aux événements](/learn/responding-to-events) pour commencer à lire ce chapitre page par page !

Reviewer: ```suggestion
Allez sur [Réagir aux événements](/learn/responding-to-events) pour commencer à lire ce chapitre page par page !
```
